File: src/packages/agent_core/builtin_tools/web_search/searxng.py
from __future__ import annotations

from typing import Any, Mapping

import httpx

from .provider import WebSearchResult


class SearxngSearchHttpError(RuntimeError):
    def __init__(self, *, status_code: int, message: str) -> None:
        super().__init__(message)
        self.status_code = status_code


class SearxngWebSearchProvider:
    def __init__(self, *, base_url: str, client: httpx.AsyncClient | None = None) -> None:
        self._base_url = base_url.rstrip("/")
        self._client = client

    async def search(self, *, query: str, max_results: int) -> list[WebSearchResult]:
        client = self._client
        if client is None:
            timeout = httpx.Timeout(None)
            async with httpx.AsyncClient(base_url=self._base_url, timeout=timeout) as client:
                return await self._search_with_client(client=client, query=query, max_results=max_results)
        return await self._search_with_client(client=client, query=query, max_results=max_results)

    async def _search_with_client(
        self,
        *,
        client: httpx.AsyncClient,
        query: str,
        max_results: int,
    ) -> list[WebSearchResult]:
        resp = await client.get(
            "/search",
            params={
                "q": query,
                "format": "json",
            },
        )
        if resp.status_code != 200:
            raise SearxngSearchHttpError(status_code=int(resp.status_code), message=f"SearXNG 返回 {resp.status_code}")

        payload = resp.json()
        if not isinstance(payload, Mapping):
            raise ValueError("SearXNG JSON 响应必须为对象")
        raw_results = payload.get("results")
        if not isinstance(raw_results, list):
            raise ValueError("SearXNG JSON 响应 results 必须为数组")

        results: list[WebSearchResult] = []
        for item in raw_results:
            if len(results) >= max_results:
                break
            if not isinstance(item, Mapping):
                continue

            title = _as_non_empty_str(item.get("title"))
            url = _as_non_empty_str(item.get("url"))
            snippet = _as_str(item.get("content")) or _as_str(item.get("snippet")) or ""
            if title is None or url is None:
                continue
            results.append(WebSearchResult(title=title, url=url, snippet=snippet))

        return results


def _as_str(value: Any) -> str | None:
    if isinstance(value, str):
        cleaned = value.strip()
        return cleaned if cleaned else None
    return None


def _as_non_empty_str(value: Any) -> str | None:
    return _as_str(value)


__all__ = ["SearxngSearchHttpError", "SearxngWebSearchProvider"]

File: src/services/worker/internal/tools/builtin/web_search/searxng.go
package websearch

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type SearxngProvider struct {
	baseURL string
	client  *http.Client
}

func NewSearxngProvider(baseURL string) *SearxngProvider {
	cleaned := strings.TrimRight(strings.TrimSpace(baseURL), "/")
	return &SearxngProvider{
		baseURL: cleaned,
		client:  &http.Client{Timeout: 15 * time.Second},
	}
}

func (p *SearxngProvider) Search(ctx context.Context, query string, maxResults int) ([]Result, error) {
	if strings.TrimSpace(query) == "" {
		return nil, fmt.Errorf("query 不能为空")
	}
	if maxResults <= 0 {
		return nil, fmt.Errorf("maxResults 必须为正整数")
	}

	parsed, err := url.Parse(p.baseURL + "/search")
	if err != nil {
		return nil, err
	}
	params := parsed.Query()
	params.Set("q", query)
	params.Set("format", "json")
	parsed.RawQuery = params.Encode()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, parsed.String(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := p.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(io.LimitReader(resp.Body, 1024*1024))
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, HttpError{StatusCode: resp.StatusCode}
	}

	var parsedJSON any
	if err := json.Unmarshal(body, &parsedJSON); err != nil {
		return nil, err
	}
	root, ok := parsedJSON.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("searxng response 类型错误")
	}

	rawResults, ok := root["results"].([]any)
	if !ok {
		return nil, fmt.Errorf("searxng response 缺少 results")
	}

	out := make([]Result, 0, maxResults)
	for _, item := range rawResults {
		if len(out) >= maxResults {
			break
		}
		obj, ok := item.(map[string]any)
		if !ok {
			continue
		}
		title, _ := obj["title"].(string)
		urlText, _ := obj["url"].(string)
		snippet, _ := obj["content"].(string)
		title = strings.TrimSpace(title)
		urlText = strings.TrimSpace(urlText)
		snippet = strings.TrimSpace(snippet)
		if title == "" || urlText == "" {
			continue
		}
		out = append(out, Result{
			Title:   title,
			URL:     urlText,
			Snippet: snippet,
		})
	}
	return out, nil
}

